### ISSUE TITLE
Mutable API

### ISSUE THREAD

**Issue by ivan-kleshnin** [selfrefactor/rambda#44]
> Rambda's `reverse` modifies the array, instead of returning reversed copy of it.

Just wanted to ask why? When you create a new reversed array it's probably just a shallow copy. There's no need to "fight" for RAM too much as JS **already** has memory sharing built-in. It's just not protected sharing and we protect it by convention with Ramda approach.

```js
let xs = [{...}, {...}, {...}]
//
let ys = [xs[2], xs[1], xs[0]]
```

`ys` does not double the memory usage. It's only additional pointers (4 bytes each)!

I.m.o the convention of "every function here is immutable" has to be kept at any cost.

**Comment by selfrefactor** [comment]
I agree with your reasoning and I will change that mutable behavior.

Initially I added `R.reverse` because I created helper method, that directly uses the native method(which mutates the array).

Methods `toString`, `toLower`, `toUpper` and `trim` uses the same helper. 

I will post again, once this issue is solved.

**Comment by selfrefactor** [comment]
The bug is fixed in the new version `0.9.6`

**Comment by ivan-kleshnin** [comment]
Great! 

Btw. Ramda adds ES6 modules in the 0.25 version so Webpack / Rollup treeshaking can be applied. Then only functions that are really used can be bundled.

So the purpose of this repo as *tinier Ramda* can be argued as "worthless" by some people.
As for me – I think the purpose of having "easier Ramda, with no fuss" is still valid. Maybe you have (or will have) some additional motivations to modify Ramda. It's certanly not the only library doing and I can totally imagine myself using one.

P.S

I would rather drop `assoc` and `assocPath` than lenses. It's a shame Ramda lenses have inconvenient API.

**Comment by selfrefactor** [comment]
Thanks for the feedback.

I didn't know that Ramda is now tree-shakeable and you are right that Rambda's argumentation should be changed to reflect this news. If you particular idea for the description of Rambda, please feel free to suggest it as I consider "easier/faster Ramda, with no fuss" more like a general guidance.

> R.id alias for R.identity
This seems valid suggestion - will do it.

Can you just explain a bit further what you mean by: 

> I would rather drop assoc and assocPath than lenses. It's a shame Ramda lenses have inconvenient API.

**Comment by ivan-kleshnin** [comment]
> BTW Which library you are using for tree-shaking?

I didn't try it yet, but I plan to. So I have no idea if it will be easy or *not-so-easy* addition.
I'll try to remember to leave a note here after that.

---

> Can you just explain a bit further what you mean by:

>> I would rather drop assoc and assocPath than lenses. It's a shame Ramda lenses have inconvenient API.

I use the following helpers:

```js
let lensify = (lens) => {
  if (lens instanceof Array) {
    return reduce(
      (z, s) => compose(z, typeof s == "number" ? lensIndex(s) : lensProp(s)),
      id,
      lens
    )
  } else if (lens instanceof Function) {
    return lens
  } else {
    throw Error(`invalid lens ${lens}`)
  }
}

// Changing global namespace for brevity (bad for libs, ok for apps)
window.R.viewL = curry((lens, obj) => view(lensify(lens), obj))
window.R.setL = curry((lens, val, obj) => set(lensify(lens), val, obj))
window.R.overL = curry((lens, fn, obj) => over(lensify(lens), fn, obj))
```

Then:

```js
R.assoc("foo", foo, {})
=>
R.setL(["foo"], foo, {}) 

// Btw, setL API can be overloaded to support strings or string masks
// as you did HERE with assocPath. I'm ambiguos about that change for now.

R.assocPath(["foo", "bar"], "fooBar", {foo: {}})
=>
R.setL(["foo", "bar"], "fooBar", {foo: {}})
```

But lenses are much more powerful because you can 

a) modify arrays

```js
R.setL(["foo", "bar", 0], "hi", {foo: {bar: []}})
```

b) apply functions with `over`

```js
R.overL(["foo", "bar", 0], (xs) => xs.concat("hi"),{foo: {bar: []}})
```

c) get items 

```
R.getL(["foo", "bar", 0], {foo: {bar: ["baz"]}}) // "baz"
```

d) make your own lenses working with anything: from bit registers to ImmutableJS objects

*No time to cover*.

So it's lesser API surface for the extended functionality.

**Comment by ivan-kleshnin** [comment]
> If you particular idea for the description of Rambda, please feel free to suggest it as I consider "easier/faster Ramda, with no fuss" more like a general guidance.

Good question. To begin with, I would not copy the equivalent API – it's a lot of work to support.
(especially for Rambda and RambdaX) 

I would just made a list of links to Ramda's original API + the difference.

**Comment by selfrefactor** [comment]
Wow- that was a long explanation and I appreciate it. 

I am not big on lenses, so I needed some time to comprehend your explanation. Still, I don't feel fully comfortable adding those methods by myself, as I have never used lenses before. I would accept PR with your suggestions, if you feel like it doing it.

As for tree-shaking - I am asking as my attempts with Rollup and Webpack are not as smooth as expected. Will keep the issue open to wait for your experience on that.

As for you latest comment - I agree that it is a lot of work supporting different method explanation, but I see it also as necessary difference between `Ramda` and `Rambda`. Also, linking to Ramda original API was actually the very first documentation solution, but the time investment in creating new examples and new method descriptions now is too large to throw it out.

**Comment by ivan-kleshnin** [comment]
> As for you latest comment - I agree that it is a lot of work supporting different method explanation, but I see it also as necessary difference between Ramda and Rambda. Also, linking to Ramda original API was actually the very first documentation solution, but the time investment in creating new examples and new method descriptions now is too large to throw it out.

Easier examples are also fine. Some of Ramda's originals are hard to follow.

> I am not big on lenses, so I needed some time to comprehend your explanation. Still, I don't feel fully comfortable adding those methods by myself, as I have never used lenses before. I would accept PR with your suggestions, if you feel like it doing it.

Ok, I'll consider that 👍

**Comment by ivan-kleshnin** [comment]
I've succeeded with working treeshaking with Ramda. I'm going to post repo examples soon. 
Two important notes:

1) Bundle size is still big (bigger than Rambda). A lot of currying machinery is pulled for every function in Ramda.

2) Treeshaking works only for production builds (probably because it's very expensive in CPU). So the only way to keep apps with Ramda fast is to vendor-split it.  Bundle size still matters **a lot**.

As for the API surface question – I keep collecting the stats on functions I actually use. 

I don't use many of `fn` combinators.

```js
var average = R.converge(R.divide, [R.sum, R.length])
average([1, 2, 3, 4, 5, 6, 7]) //=> 4
```

vs

```
let average = (xs) => reduce(R.add, 0, xs) / xs.length
average([1, 2, 3, 4, 5, 6, 7]) //=> 4
```

I'm yet to find cases where `converge`-level abstractions are easier to read and write then expanded lambdas.

**Comment by ivan-kleshnin** [comment]
```js
+ map2 = addIndex(map)
- converge
+ filter2 = addIndex(filter)
- update // replace with lenses
- adjust // replace with lenses
- assoc // replace with lenses
- assocPath // replace with lenses
- applySpec // overthinked
...
```

**Comment by selfrefactor** [comment]
> I've succeeded with working treeshaking with Ramda. I'm going to post repo examples soon. 

Really nice. I also played further with tree-shaking and I found out that I need to remove some of Rambda internal helpers, as Rollup otherwise struggles to perform tree-shake. With `Ramda` and I could create working tree-shaking settings as their helpers also create obstacle for Rollup, so I am very curios how you made that. I will wait for your repo to be ready. 

> A lot of currying machinery is pulled for every function in Ramda.

One of the reason for me to build my own `Ramda` - too much unnecessary magic.

---
As for your suggestions for Rambda - I agree with you that those are required in order to make the library more serious.

**Comment by ivan-kleshnin** [comment]
> With Ramda and I could create working tree-shaking settings as their helpers also create obstacle for Rollup, so I am very curios how you made that. I will wait for your repo to be ready.

I used Webpack. I never tried Rollup – learning both is too expensive in time, unfortunately.

**Comment by selfrefactor** [comment]
I am posting a link to a repo, which prove that `Rambda` is more tree-shakeable than `Ramda`

[https://github.com/selfrefactor/tree-shaking-example](https://github.com/selfrefactor/tree-shaking-example)

**Comment by ivan-kleshnin** [comment]
Thank you. Check also this https://github.com/scabbiaza/ramda-webpack-tree-shaking-examples/pull/2 **output bundle - 877 B**. Looks like there is some magic Webpack plugin that enables a normal tree-shaking :)

**Comment by selfrefactor** [comment]
@ivan-kleshnin It was nice having this issue-chat but I will close it, as it seems we run out of steam :)
